The trying module
=================

    >>> from pysyte.testing import trying
    >>> assert 'running of doctests' in trying.__doc__

Imports for tests
-----------------

    >>> import os
    >>> import getpass

Trying trying
-------------

Following tests will only work when run under trying itself
    because, e.g. it provides a global called test
    >>> try:
    ...     assert test
    ... except NameError:
    ...     raise KeyboardInterrupt

Running tests with try provides an instance called test
    which has some interesting attributes
    e.g. this file is called try.tests
    >>> assert test.path.name == 'trying.test'

If that went well we have a hidden variable called test
    which is an object from the __main__ module
    and has some attributes

    >>> assert test.path_to_py.name == 'trying.py'

    >>> assert test.here == os.getcwd()

We also have access to some modules I often use
    without a need for explicit import
    Which ignores the heuristic that
        "Explicit is better than implicit"
    >>> assert test.path.name == makepath(sys.argv[0]).name

And we can run bash commands
    >>> assert bash('/bin/ls /bin/ls')
    /bin/ls

    >>> assert test.home == makepath('~')

We know the user who is running the tests
    >>> assert test.user == '%s@%s' % (getpass.getuser(), test.host)

Edge cases
----------

    >>> assert not bash('/bin/ls /no_such_file')
    FAIL: ...No such file or directory

    >>> show(os)
    Help on module os:
    ...

    >>> show(0)
    0

    >>> import see
    >>> assert see == trying.make_module(test.path_to_py.parent / 'see.py')
    >>> show(see)
    Help on module see:
    ...
        /.../testing/see.py
    ...

    >>> del sys.modules['see']
    >>> see = trying.make_module(test.path_to_py.parent / 'see.py')
    >>> show(see)
    Help on package see:
    ...

    >>> mod = trying.make_module(test.path_to_py.parent / 'not_a_file.py')
    >>> assert isinstance(mod, type(trying))
    >>> assert mod.__name__, 'not_a_file'
